### PR TITLE
Added api exeptions

### LINKED ISSUE #431: As a developer, i would like to have Error handling when doing API calls from BloCs, as to make the weekplanner more responsive
**Is your feature request related to a problem? Please describe.**
As of now A lot of the BloCs are not handling exceptions/errors thrown when doing an API call. 
This should be handled, and the response should be sent to the screen, with a response message to that error. 
This is now possible, regarding https://github.com/aau-giraf/weekplanner/issues/349. ( is in a comment)
This would make the App better, as it is more responsive when doing requests. 

**Describe the solution you'd like**
I would like to have whoever is on this issue, to create other issues as https://github.com/aau-giraf/weekplanner/issues/429, https://github.com/aau-giraf/weekplanner/issues/428 to all the places, where API calls are made, but no error/Exception handling is made.

**Describe alternatives you've considered**
I do not consider any alternative solutions. Error handling is Perile for a good application.

### PR DESCRIPTION
**Pull request by Emil-Linnebjerg**
This adds api exeptions to the bloc.
This is the result of issue #431 
There has been created exeption handeling for api calls in the bloc.

### CODE REVIEW

**Comment by Determinatoro** [comment]
Please use the following checklist, @Oliversrensen. Feel free to check it off as you go along.

**Code Design** 

- [x] The code is in the right place? (Both in terms of folder structure and class structure)
- [x] The code is not over-engineered. Examples of over-engineering: Implemented behavior for future problems     
- [x] This code could not, to the best of my knowledge, have reused existing code
- [x] The code does not introduce functionality that is unnecessary for solving the problem

**Code Readability** 

- [x] Names are meaningful and self-documenting
- [x] It is understandable, by reading the code, what it does
- [ ] The code is simple and readable. i.e. it contains no \"hack\", or obscure solution   

**Code Maintainability**

- [x] Has all the added code been properly commented

Only check one of the items below:

- [ ] The functionality does **not** need to be covered by tests
- [x] The functionality needs to be covered by tests

Only check these if the functionality needs to be covered by tests:

- [ ] All functionality is covered by tests
- [ ] The tests has sensible names
- [ ] By reading the tests you know what they cover
- [ ] The tests cover sensible cases. Both happy paths and exception paths
- [ ] The tests cover the full functionality. i.e., The tests fails if some of the requested functionality is missing

**Functionality**

For this part, open the app, and test:
- [ ] The features that the code claims to implement, are actually implemented
- [x] The code does not look like it has bugs
- [ ] All the new screens are reachable through in-app navigation

**Comment by Determinatoro** [comment]
Please use the following checklist, @ChristofferAaen. Feel free to check it off as you go along.

**Code Design** 

- [x] The code is in the right place? (Both in terms of folder structure and class structure)
- [x] The code is not over-engineered. Examples of over-engineering: Implemented behavior for future problems     
- [x] This code could not, to the best of my knowledge, have reused existing code
- [x] The code does not introduce functionality that is unnecessary for solving the problem

**Code Readability** 

- [x] Names are meaningful and self-documenting
- [x] It is understandable, by reading the code, what it does
- [x] The code is simple and readable. i.e. it contains no \"hack\", or obscure solution   

**Code Maintainability**

- [x] Has all the added code been properly commented

Only check one of the items below:

- [ ] The functionality does **not** need to be covered by tests
- [x] The functionality needs to be covered by tests

Only check these if the functionality needs to be covered by tests:

- [x] All functionality is covered by tests
- [x] The tests has sensible names
- [ ] By reading the tests you know what they cover
- [ ] The tests cover sensible cases. Both happy paths and exception paths
- [ ] The tests cover the full functionality. i.e., The tests fails if some of the requested functionality is missing

**Functionality**

For this part, open the app, and test:
- [ ] The features that the code claims to implement, are actually implemented
- [ ] The code does not look like it has bugs
- [ ] All the new screens are reachable through in-app navigation

**Review by OSanderDK — DISMISSED** [review]

**Review by ChristofferAaen — DISMISSED** [review]
Okay I see

**Comment by LivHolm** [comment]
@Emil-Linnebjerg Det er dit job at resolve de merge konflikter der er.

**Review by OSanderDK — DISMISSED** [review]

**Review by ChristofferAaen — APPROVED** [review]

**Review by ChristofferAaen — CHANGES_REQUESTED** [review]
Det ser ud til du skriver error codes ud i consolen. Inde i API mappen i weekplanner med en error translater

**Review by ChristofferAaen — CHANGES_REQUESTED** [review]
I fanger ikke nogle API exceptions 
brug den her:
https://github.com/aau-giraf/weekplanner/blob/develop/lib/api/errorcode_translater.dart

**Comment by Emil-Linnebjerg** [comment]
Jeg er ikke helt sikker på jeg forstår vi har testet det den fanger fejl

**Comment by ChristofferAaen** [comment]
https://github.com/aau-giraf/api_client/blob/master/lib/api/api_exception.dart
er den type af exceptions i skal fange
https://github.com/aau-giraf/weekplanner/blob/develop/lib/api/errorcode_translater.dart
oversætter exceptions så brugeren kan forstå dem

Disse klasser fanger exceptions og oversætter dem til dansk, således brugeren kan forstå dem. I fanger ikke nogen af disse exceptions, hvilket er det i skal. (som nævnt til sprint 2 review)

**Comment by Emil-Linnebjerg** [comment]
Jeg kan godt se i har lavet en metode til at smide fejlt det super men den kræver en "context" som man får når man er i screens
Api fejl skal smides i blocs og derfor har vi ingen "context" 
Denne metode der nu kommer til api er lavet til at teste på api kald i blocs

File: lib/screens/upload_image_from_phone_screen.dart [hunk 1]
```diff
@@ -157,22 +151,33 @@ class UploadImageFromPhone extends StatelessWidget {
...
+          ),
+          color: theme.GirafColors.white70,
+          borderRadius: _imageBorder),
+      child: Column(
+        mainAxisAlignment: MainAxisAlignment.center,
+        crossAxisAlignment: CrossAxisAlignment.center,
+        children: <Widget>[
+          Image.asset(
+            'assets/icons/gallery.png',
+            color: theme.GirafColors.black,
```
Comment (ChristofferAaen) [inline lib/screens/upload_image_from_phone_screen.dart]: what is this??? er det en del af jeres issue??

**Comment by Emil-Linnebjerg** [comment]
"As of now A lot of the BloCs are not handling exceptions/errors thrown when doing an API call."
Jeg tror desværre i har misforstået den opgave der skal løses her.

**Comment by ChristofferAaen** [comment]
I bruger også try, catch på det meste, hvilket også står i issuet man ikke skal. I er nød til at bruge onError, fra api exceptions. 

Er meget ked af det men tror det hele skal laves om forfra.

**Comment by ChristofferAaen** [comment]
fordi i ikke får fanget de rigtige exceptions fra API

**Review by ChristofferAaen — CHANGES_REQUESTED** [review]

**Comment by ChristofferAaen** [comment]
https://github.com/aau-giraf/wiki/blob/develop/docs/development/WeekplannerErrorHandlingGuidelines.md

Process fortalte det her til sprint review sidst

**Comment by Emil-Linnebjerg** [comment]
Så vi bruger try{}on IKKE try catch 
Det gør at vi kan få en specifik fejl som man skal gøre dette gør man får at give en app som er responsive med specifikke fejl istedet for generiske fejl der ikke hjælper brugeren.
Eksempelvis en SocketException smider en fjel der promter brugeren med en besked der læser "kunne ikke få forbindelse til internettet" så ved brugeren præcis hvad problemet er og synes ikke at appen lagger eller er buggy.
Til koden er det her den nye version af weekplaner for at undgå merge conflicts.
Er glad for i vil se gode exeptions men tror stadig i misforstår problemet der skal løses her.

**Comment by ChristofferAaen** [comment]
Kan i ikke tage at få dækket linjerne i jeres kode med test, fordi CI'en fejler. 

Vi kan godt se i fanger nogle exceptions, men kan i så ikke tilføje en onError på streams, vi kan se i har en context som i siger i ikke kan få fat i, kan i ikke bare bruge den. 

Men vigtigt i lige kan få dem igennem CI'en

**Comment by Emil-Linnebjerg** [comment]
Vi er i gang med at gøre præcis det tænker det er fint hellere lidt flere test en at have for få

**Comment by Emil-Linnebjerg** [comment]
Ok håber alle er glade nu. Vi fanger exceptions, hvis der kommer en error fanger vi også den tager den så igennem exception handleren og sender den til jeres error handler med en context. Altså er både exceptions og errors fanget og håndteret nu.

**Comment by ChristofferAaen** [comment]
kan også se at mange linjer ikke er dækket af test. Er det muligt? Process mener også det skal testet

**Comment by Emil-Linnebjerg** [comment]
Ok testen er blevet opdateret nu så codecov kan forstå den. Det burde dække den kode der giver mening at teste.

**Comment by LivHolm** [comment]
@Emil-Linnebjerg husk at resolve merge konflikter

**Comment by OSanderDK** [comment]
Jeg er ved at kigge på det

File: lib/blocs/blocs_api_exeptions.dart [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+///This class handles all api exeptions
+class BlocsApiExeptions{
+
+
+  ///this constructor allows developers to costumize errors
+  BlocsApiExeptions(String errorType,
+      [this.errormessage = 'opps noget gik galt',
+        Object error]){
+    ///switchen her er bare for at give nogle basale api fejl
```
Comment (OSanderDK) [inline lib/blocs/blocs_api_exeptions.dart]: Der er skrevet code comments på dansk, de skal gerne være på engelsk. Gælder flere steder.

File: lib/blocs/blocs_api_exeptions.dart [hunk 3]
```diff
@@ -0,0 +1,70 @@
+import 'package:flutter/cupertino.dart';
+import 'package:flutter/material.dart';
+import 'package:weekplanner/api/errorcode_translater.dart';
+import '../main.dart';
+
+///This class handles all api exeptions
+class BlocsApiExeptions{
```
Comment (OSanderDK) [inline lib/blocs/blocs_api_exeptions.dart]: Exception er stavet forkert i klassens navn. "Exeption" skal være "Exception".

File: lib/blocs/blocs_api_exeptions.dart [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+  String errormessage;
+
+  ///dette gøre man bare kan kaste vores api fejl
+ @override
+  String toString() {
+   return errormessage;
+ }
+
+ ///denne del lave en alret besked som popper op til brugeren
+  void showMyDialog() {
```
Comment (OSanderDK) [inline lib/blocs/blocs_api_exeptions.dart]: Der kommer ikke en AlertDialog.

File: lib/blocs/blocs_api_exeptions.dart [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    showDialog<Center>(
+        context: navigatorKey.currentContext,
+        builder: (BuildContext context) {
+          return AlertDialog(
```
Comment (OSanderDK) [inline lib/blocs/blocs_api_exeptions.dart]: Designguiden kræver at man bruger GirafNotifyDialog og ikke AlertDialog :)

**Review by OSanderDK — CHANGES_REQUESTED** [review]

File: lib/blocs/blocs_api_exeptions.dart (hunk 5 above)
Comment (Emil-Linnebjerg) [reply lib/blocs/blocs_api_exeptions.dart]: True stuff

File: lib/blocs/blocs_api_exceptions.dart [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    }
+  }
+
+  final ApiErrorTranslater _translator = ApiErrorTranslater();
+
+  ///This string is used to contain error messages
+  String errormessage;
+
+ ///This part creates the message
+  void showMyDialog() {
```
Comment (OSanderDK) [inline lib/blocs/blocs_api_exceptions.dart:45]: Har du selv set om du kan fremtvinge en popup? Vi kan ikke få det til at virke 👎

**Review by OSanderDK — CHANGES_REQUESTED** [review]
Hvor blev dine tests af? 😕

**Review by OSanderDK — CHANGES_REQUESTED** [review]
Jeg vil gerne henvise dig til mit tidligere review, hvor jeg er i tvivl om det overhovedet virker, og undrer mig over hvor dine tests er blevet af.

**Pull request closed without merging**